### PR TITLE
[16.0][ADD] delivery_fee: new module

### PR DESCRIPTION
**Pull request by chienandalu**
https://www.loom.com/share/93fad82c79284f2a83fa816955164763

cc @moduon MT-13311

### CODE REVIEW

File: delivery_fee/reports/delivery_slip_report.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<odoo>
+    <template
+        id="report_delivery_document2"
+        inherit_id="delivery.report_delivery_document2"
+    >
+        <xpath expr="//div[hasclass('page')]" position="inside">
+            <t t-set="fee_product" t-value="o.carrier_id.fee_product_id" />
```
Comment (EmilioPascual) [inline delivery_fee/reports/delivery_slip_report.xml]: Perhaps you could create a template and use it in both reports. They are 99% identical. In any case, it works the same way.

**Review by EmilioPascual — APPROVED** [review]
Wonderful module. A minor comment, in no case blocking

File: delivery_fee/reports/delivery_slip_report.xml (hunk 1 above)
Comment (chienandalu) [reply delivery_fee/reports/delivery_slip_report.xml]: Yep, I can do it :)

**Review by antoniodavid — APPROVED** [review]
LGTM

**Review by Andrii9090-tecnativa — APPROVED** [review]
Code & funcional review 👍 

Image

**Review by Shide — APPROVED** [review]

**Review by marcos-mendez — CHANGES_REQUESTED** [review]
**Automated Review -- Tests Failed**

### 1. Root Cause of the Test Failure

The test failure is due to a **database connection error**, not a code issue in the module itself. The error occurs during Odoo startup (`odoo.cli.main()`), likely because the test environment is misconfigured or the database specified in the connection string is unreachable.

### 2. Suggested Fix

No code fix is needed for the module. However, to resolve the test failure:

- Ensure the database URL in the test configuration is correct.
- Confirm that the PostgreSQL service is running and accessible at `postgres.runboat-infra.svc.cluster.local:5432`.
- Verify that the test database `runboat@postgres.runboat-infra.svc.cluster.local:5432` exists and is accessible.

### 3. Additional Code Issues

#### a. **Potential NPE in `delivery_slip_report.xml`**
In `delivery_fee/reports/delivery_slip_report.xml`, line ~13:

```xml
<t t-set="fee_line" t-value="o.sale_id.order_line.filtered(lambda x: x.is_delivery_fee and x.delivery_fee_picking_id == o)" />
```

If `o.sale_id` is `False` (e.g., for internal pickings), this could cause an AttributeError. It should be guarded:

```xml
<t t-if="o.sale_id" t-set="fee_line" t-value="o.sale_id.order_line.filtered(lambda x: x.is_delivery_fee and x.delivery_fee_picking_id == o)" />
```

#### b. **Missing `default=False` in `is_delivery_fee` field**
In `delivery_fee/models/sale_order.py`, line 94:

```python
is_delivery_fee = fields.Boolean(string="Is a Delivery Fee", default=False)
```

This is correct, but ensure that this field is properly initialized in all contexts where it's used.

#### c. **Inconsistent naming in README**
In `delivery_fee/README.rst` and `readme/USAGE.md`, the phrase "delivey fees" is misspelled. Should be "delivery fees".

### 4. Test Improvements

To better cover this module, add tests in the following categories:

#### a. **TransactionCase Tests**
Use `TransactionCase` for testing core logic:

- **Test fee creation on delivery**: Create a sale order with a delivery carrier that has a fee product, confirm the order, and check if a fee line was added.
- **Test fee exemption**: Create a partner with `delivery_fee_exemption=True` and verify no fee line is added.
- **Test return handling**: Simulate a full return of a picking and verify that the fee is refunded (based on `fee_return_percentage`).

#### b. **SavepointCase Tests**
Use `SavepointCase` for more complex scenarios involving data cleanup:

- **Test `one_delivery_fee_by_sale_order` setting**: Enable the setting and confirm that only one fee line is created even if multiple outgoing pickings are done.
- **Test fiscal position mapping**: Verify that taxes are correctly mapped when fiscal positions are applied.

#### c. **Tagged Tests**
Use OCA test tags like `@tag('delivery_fee')` to group tests and make them easy to run.

Example test structure:
```python
from odoo.tests import TransactionCase

class TestDeliveryFee(TransactionCase):
    def setUp(self):
        super().setUp()
        self.partner = self.env['res.partner'].create({'name': 'Test Partner'})
        self.product = self.env['product.product'].create({
            'name': 'Fee Product',
            'type': 'service',
            'list_price': 10.0,
        })
        self.carrier = self.env['delivery.carrier'].create({
            'name': 'Test Carrier',
            'fee_product_id': self.product.id,
        })
        self.sale_order = self.env['sale.order'].create({
            'partner_id': self.partner.id,
            'carrier_id': self.carrier.id,
        })

    def test_create_fee_line(self):
        # Test logic here
```

This ensures that edge cases and business logic are thoroughly tested.

---
⏰ This PR has been open for **52 days**.

> _Every ignored PR is a contributor who might not come back. Review time matters. ([OCA Aging Report](https://git.pop.coop/pop/oca-reviewer))_

---
### Reciprocal Review Request



Hi everyone! I found some test failures on this PR and left detailed feedback above. I am happy to discuss or help debug. In the meantime, if any of you get a chance, I would appreciate a look at my open PR(s):

**My open PRs across OCA:**
- [server-tools#3548 [18.0][MIG] base_kanban_stage: Migration to 18.0](https://github.com/OCA/server-tools/pull/3548)
- [hr-attendance#262 [16.0][ADD] Hr_attendance_idsecure: iDSecure (ControliD) attendance integration](https://github.com/OCA/hr-attendance/pull/262)
- [stock-logistics-workflow#2276 [16.0][ADD] stock_move_line_devaluation](https://github.com/OCA/stock-logistics-workflow/pull/2276)
- [stock-logistics-workflow#2275 [16.0][ADD] Stock move line analytic account](https://github.com/OCA/stock-logistics-workflow/pull/2275)
- [stock-logistics-workflow#2268 [16.0][ADD] stock_move_line_picking_partner](https://github.com/OCA/stock-logistics-workflow/pull/2268)
- [purchase-workflow#2694 [16.0][IMP]Purchase workflow added to review state & exception fix](https://github.com/OCA/purchase-workflow/pull/2694)

Reviewing each other's work helps the whole community move forward. Thank you!

---
_Environment via [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer): Minikube + K8s Job + oca-ci/py3.10-odoo16.0 | Odoo 16.0_
_Automated review by [OCA Neural Reviewer](https://git.pop.coop/pop/oca-reviewer) + qwen3-coder:30b_

**Comment by Shide** [comment]
@marcos-mendez your last AI review is totally unnecessary, please keep comments clear and useful.
If all of us use AI like this, comments history would be irrelevant.

**Review by marcos-mendez — APPROVED** [review]
